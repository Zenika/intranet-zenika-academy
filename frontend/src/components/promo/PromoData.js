import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import Axios from 'axios';
import Moment from 'react-moment';
import './PromoData.scss';

class PromoDetails extends Component {
  constructor(props) {
    super(props);
    this.state = {
      users: [],
      program: {},
      promotion: {},
      isAdmin: false,
      redirectToAdmin: false,
      redirectToPromoEdit: false,
    };
    this.handleDelete = this.handleDelete.bind(this);
    this.handleDeleteClick = this.handleDeleteClick.bind(this);
    this.handleDeleteEnterKey = this.handleDeleteEnterKey.bind(this);
    this.goToPromoEdit = this.goToPromoEdit.bind(this);
  }

  componentDidMount() {
    document.title = 'Détails de la promo';
    const role = sessionStorage.getItem('userRole');
    if (JSON.parse(role) === 1) {
      this.setState({ isAdmin: true });
    }
    const { match } = this.props;
    const url = `/api/promotions/details/${parseInt(match.params.id, 10)}`;
    Axios.get(url).then((result) => {
      this.setState({
        users: result.data.users,
        program: result.data.program,
        promotion: result.data.promotion,
      });
    });
  }

  /**
   * Allows to delete a promo in the DB
   * delete students
   * update teachers with promoId null
   * @param {*} id Promo Id
   */
  handleDelete(id) {
    const url = `/api/promotions/${id}`;
    // eslint-disable-next-line no-restricted-globals,no-alert
    if (confirm('Voulez vous supprimer cette promotion?')) {
      Axios.delete(url)
        .then(() => {
          this.setState({ redirectToAdmin: true });
        })
        .catch((err) => err);
    }
  }

  /**
   * Allows to call the delete method on mouse click
   * @param {*} id Promo Id
   */
  handleDeleteClick(id) {
    return this.handleDelete(id);
  }

  /**
   * Allows to call the delete method on enter Key press
   * @param {*} id Promo Id
   */
  handleDeleteEnterKey(e, id) {
    if (e.key === 'Enter') {
      return this.handleDelete(id);
    }
    return false;
  }

  goToPromoEdit() {
    return this.setState({ redirectToPromoEdit: true });
  }

  render() {
    const {
      users,
      program,
      promotion,
      redirectToAdmin,
      isAdmin,
      redirectToPromoEdit,
    } = this.state;
    const { handleDeleteClick, handleDeleteEnterKey, goToPromoEdit } = this;
    const teachers = users.filter((user) => user.role === 2);
    const students = users.filter((user) => user.role === 3);
    const programTitle = program ? program.title : 'Pas de programme associé';
    const detailLink = program ? (
      <a
        href={`/program/${program.id}/details`}
        title="Détails du programme"
        className="detailsLink"
      >
        détails...
      </a>
    ) : null;

    if (redirectToAdmin) {
      return <Redirect to="/home/admin" />;
    }

    if (redirectToPromoEdit) {
      const path = `/admin/promo/edit/${promotion.id}`;
      return (
        <Redirect
          to={{
            pathname: path,
            data: { teachers, students, promotion },
          }}
        />
      );
    }
    return (
      <>
        <div className="container">
          <div className="PromoDataHeader">
            <div className="promoInfo">
              <h1 className="title is-1 promoTitle">{promotion.title}</h1>
              <h2 className="subtitle is-3">
                à {promotion.city} du{' '}
                <Moment format="DD/MM/YYYY">{promotion.startDate}</Moment> au{' '}
                <Moment format="DD/MM/YYYY">{promotion.endDate}</Moment>
              </h2>
            </div>
            {isAdmin && (
              <div className="buttonContainer">
                <a
                  href="/home/admin"
                  className="button is-dark goBack"
                  title="Revenir à l'accueil"
                >
                  Revenir à l&apos;accueil
                </a>
                <button
                  title=" Edition d'une promo"
                  type="button"
                  className="button is-warning"
                  onClick={goToPromoEdit}
                >
                  Editer
                </button>
                <button
                  title="Suppression d'une promo"
                  type="button"
                  className="button is-danger"
                  onClick={() => handleDeleteClick(promotion.id)}
                  onKeyUp={() => handleDeleteEnterKey(promotion.id)}
                >
                  Supprimer
                </button>
              </div>
            )}
          </div>
          <div className="container">
            <div className="notification">
              <h3 className="title is-3">Formateurs</h3>
              <ul className="promoTeachers">
                {teachers.map((teacher) => (
                  <li
                    key={`teacher-${teacher.id}`}
                  >{`${teacher.firstName} ${teacher.lastName}`}</li>
                ))}
              </ul>
            </div>
          </div>
          <div className="container">
            <div className="notification">
              <h3 className="title is-3">Elèves</h3>
              <ul className="promoStudents">
                {students.map((student) => (
                  <li
                    key={`student-${student.id}`}
                  >{`${student.firstName} ${student.lastName}`}</li>
                ))}
              </ul>
            </div>
          </div>
          <div className="container">
            <div className="notification">
              <h3 className="title is-3">Programme</h3>
              <p>{programTitle}</p>
              {detailLink}
            </div>
          </div>
        </div>
      </>
    );
  }
}

export default PromoDetails;
