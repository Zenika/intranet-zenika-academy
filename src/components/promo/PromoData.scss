.notification {
  margin-bottom: 1rem;
}

.buttonContainer{
  display: flex;
  width: 30rem;
  justify-content: space-between;
}

.PromoDataHeader {
  display: flex;
  flex-flow: row nowrap;
  justify-content: space-between;

  .promoInfo {
    margin-bottom: 1rem;
  }
}