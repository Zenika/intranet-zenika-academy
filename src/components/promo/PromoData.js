import React from 'react';
import Axios from 'axios';
import Moment from 'react-moment';
import './PromoData.scss';
// import { CSSTransitionGroup } from 'react-transition-group';
// import PropTypes from 'prop-types';

// const Tab = (props) => {
//   const { name, activeTab, changeActiveTab } = props;
//   return (
//     <li className={name === activeTab && 'is-active'} onClick={() => changeActiveTab(name)}>
//       <a>
//         <span>{name}</span>
//       </a>
//     </li>
//   );
// };

// const activeTabContent = (props) => <div>{props.content}</div>;

// const tablist = [
//   {
//     name: 'Elèves',
//     content: '',
//   },
//   {
//     name: 'Formateurs',
//     content: '',
//   },
//   {
//     name: 'Programmes',
//     content: '',
//   },
// ];

// class PromoTabs extends React.Component {
//   render() {
//     const { activeTab, changeActiveTab } = this.props;
//     return (
//       <div className="tabs">
//         <ul>
//           {
//             tablist.map((tab) => (
//               <Tab
//                 tab={tab}
//                 key={tab.name}
//                 activeTab={activeTab}
//                 changeActiveTab={changeActiveTab}
//               />
//             ))
//           }
//         </ul>
//       </div>
//     );
//   }
// }

// PromoTabs.propTypes = {
//   tablist: PropTypes.shape({
//     name: PropTypes.string,
//     content: PropTypes.arrayOf(PropTypes.string),
//   }).isRequired,
//   activeTab: PropTypes.string,
//   changeActiveTab: PropTypes.func,
// };


class PromoDetails extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      users: [],
      program: {},
      promotion: {},
    };
  }

  componentDidMount() {
    const url = `http://localhost:4000/api/promotions/details/${parseInt(this.props.match.params.id, 10)}`;
    Axios.get(url)
      .then((result) => {
        this.setState({
          users: result.data.users,
          program: result.data.program,
          promotion: result.data.promotion,
        });
      });
  }

  render() {
    const { users, program, promotion } = this.state;
    const teachers = users.filter((user) => user.role === 2);
    const students = users.filter((user) => user.role === 3);
    return (
      <>
        <div className="container">
          <h1 className="title is-1">{promotion.title}</h1>
          <h2 className="subtitle is-3">
à
            {' '}
            {promotion.city}
            {' '}
            du
            {' '}
            <Moment format="DD/MM/YYYY">{promotion.startDate}</Moment>
            {' '}
            au
            {' '}
            <Moment format="DD/MM/YYYY">{promotion.endDate}</Moment>
          </h2>
          <div className="container">
            <div className="notification">
              <h3 className="title is-3">Formateurs</h3>
              {
                  teachers.map((teacher) => <li>{`${teacher.firstName} ${teacher.lastName}`}</li>)
                }
            </div>
          </div>
          <div className="container">
            <div className="notification">
              <h3 className="title is-3">Elèves</h3>

              {
                  students.map((student) => <li>{`${student.firstName} ${student.lastName}`}</li>)
                }
            </div>
          </div>
          <div className="container">
            <div className="notification">
              <h3 className="title is-3">Programme</h3>
              <p>{program.title}</p>
            </div>
          </div>
        </div>
      </>
    );
  }
}

export default PromoDetails;
