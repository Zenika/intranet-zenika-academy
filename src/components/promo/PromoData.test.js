import React from 'react';
import Axios from 'axios';
import { shallow } from 'enzyme';
import TestUtils from 'react-dom/test-utils';
import PromoData from './PromoData';

jest.mock('Axios', () => {
  const data = {
    data: {
      users: [
        {
          id: 1,
          firstName: 'Joe',
          lastName: 'McGinnis',
          email: 'mcginis@mcginis.com',
          role: 2,
          promotionId: null,
        },
        {
          id: 2,
          firstName: 'Jena',
          lastName: 'McGenty',
          email: 'mcgenty@gmail.com',
          role: 2,
          promotionId: null,
        },
        {
          id: 3,
          firstName: 'Jena',
          lastName: 'McGenty',
          email: 'mcgenty@gmail.com',
          role: 3,
          promotionId: 1,
        },
        {
          id: 4,
          firstName: 'Jena',
          lastName: 'McGenty',
          email: 'mcgenty@gmail.com',
          role: 3,
          promotionId: 1,
        },
      ],
      program: {
        id: 1,
        title: 'Consultant React',
        type: 1,
        content: [],
      },
      promotion: {
        id: 1,
        title: 'PARIS 2',
        city: 'Paris',
        startDate: '2020-10-19T00:00:00.000Z',
        endDate: '2027-01-20T00:00:00.000Z',
        programId: 1,
        createdAt: '2028-11-19T00:00:00.000Z',
        updatedAt: '2028-11-19T00:00:00.000Z',
      },
    },
  };
  return {
    get: jest.fn().mockReturnValue(Promise.resolve(data)),
  };
});

describe('Component PromoData', () => {
  const match = { params: { id: 1 } };
  const axiosMock = jest.spyOn(Axios, 'get');
  const promoData = shallow(<PromoData match={match} />);
  const promoDataInstance = promoData.instance();

  beforeEach(() => {
    promoDataInstance.componentDidMount();
  });
  test('it renders correctly', () => {
    expect(promoData).toMatchSnapshot();
  });
  test('Axios GET method is called when component is mounted', () => {
    expect(axiosMock).toBeCalled();
    expect(axiosMock).toHaveReturned();
  });
  test('Component main sections exists', () => {
    expect(promoData.find('.container')).toBeDefined();
    expect(promoData.find('.container')).toHaveLength(4);
  });
  test('Teachers section exists', () => {
    expect(promoData.find('.promoTeachers')).toBeDefined();
    expect(promoData.find('.promoTeachers')).toHaveLength(1);
  });
  test('Students section exists', () => {
    expect(promoData.find('.promoStudents')).toBeDefined();
    expect(promoData.find('.promoStudents')).toHaveLength(1);
  });
});

describe('Go back button', () => {
  const match = { params: { id: 1 } };
  const detailsPage = TestUtils.renderIntoDocument(<PromoData match={match} />);
  const goBackButton = TestUtils.findRenderedDOMComponentWithClass(detailsPage, 'goBack');

  test('should link to Admin home page', () => {
    expect(goBackButton).toBeDefined();
    expect(goBackButton.textContent).toEqual('Revenir Ã  l\'accueil');
    expect(goBackButton.getAttribute('href')).toBe('/home/admin');
  });
});
