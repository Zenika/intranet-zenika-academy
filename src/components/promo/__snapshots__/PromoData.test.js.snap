// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component PromoData it renders correctly 1`] = `
<Fragment>
  <div
    className="container"
  >
    <div
      className="PromoDataHeader"
    >
      <div
        className="promoInfo"
      >
        <h1
          className="title is-1 promoTitle"
        />
        <h2
          className="subtitle is-3"
        >
          à
           
           
          du
           
          <t
            ago={false}
            calendar={false}
            decimal={false}
            element={null}
            filter={[Function]}
            format="DD/MM/YYYY"
            fromNow={false}
            interval={60000}
            local={false}
            onChange={[Function]}
            titleFormat=""
            toNow={false}
            unit={null}
            unix={false}
            utc={false}
            withTitle={false}
          />
           
          au
           
          <t
            ago={false}
            calendar={false}
            decimal={false}
            element={null}
            filter={[Function]}
            format="DD/MM/YYYY"
            fromNow={false}
            interval={60000}
            local={false}
            onChange={[Function]}
            titleFormat=""
            toNow={false}
            unit={null}
            unix={false}
            utc={false}
            withTitle={false}
          />
        </h2>
      </div>
    </div>
    <div
      className="container"
    >
      <div
        className="notification"
      >
        <h3
          className="title is-3"
        >
          Formateurs
        </h3>
        <ul
          className="promoTeachers"
        />
      </div>
    </div>
    <div
      className="container"
    >
      <div
        className="notification"
      >
        <h3
          className="title is-3"
        >
          Elèves
        </h3>
        <ul
          className="promoStudents"
        />
      </div>
    </div>
    <div
      className="container"
    >
      <div
        className="notification"
      >
        <h3
          className="title is-3"
        >
          Programme
        </h3>
        <p />
        <a
          className="detailsLink"
          href="/home/program/details/undefined"
          title="Détails du programme"
        >
          détails...
        </a>
      </div>
    </div>
  </div>
</Fragment>
`;
